### PR TITLE
Add backend support for recovery keys

### PR DESCRIPTION
**Pull request by dsarlis**
This PR adds the necessary backend support for recovery keys -- specifically `DeviceData` is updated to include two more fields that track relevant information for keys added by the users.

### CODE REVIEW

File: src/internet_identity/src/main.rs [hunk 1]
```diff
@@ -44,11 +44,33 @@ type FrontendHostname = String;
...
+    SeedPhrase,
+}
+
 #[derive(Clone, Debug, CandidType, Deserialize)]
 struct DeviceData {
     pubkey: DeviceKey,
     alias: String,
     credential_id: Option<CredentialId>,
+    purpose: Option<Purpose>,
+    key_type: Option<KeyType>,
```
[the remark this answers is not in the document]
Comment (nomeata) [reply src/internet_identity/src/main.rs:73]: Which means we need two different rust `struct`'s, one for the storage and one for the frontend. Which is fine, I think.

[the remark this answers is not in the document]
Comment (dsarlis) [reply src/internet_identity/src/main.rs:73]: Well, I'm not sure if that's worth the trouble compared to just making them optional on the interface level.

[the remark this answers is not in the document]
Comment (dsarlis) [reply src/internet_identity/src/main.rs:73]: Hmm, ok, it doesn't look that bad but still I have a feeling we're not using Candid's flexibility at the fullest here. Anyway, I don't have very strong opinion on this, so I'll go with what you guys want.

**Comment by dsarlis** [comment]
@kritzcreek I suppose Selenium tests won't pass until the frontend is updated to use the new interface. (Joachim will help me with the backend tests, it needs some advanced Haskell knowledge it seems). 

How do you want to do this? Build on top of this PR?

**Comment by kritzcreek** [comment]
> @kritzcreek I suppose Selenium tests won't pass until the frontend is updated to use the new interface, which is another reason why it's generally a good idea to keep interfaces backwards compatible :).
> How do you want to do this? Build on top of this PR?

I can push the minimal frontend change to keep things building on this PR. The Selenium tests unfortunately scratch various timeouts so they're more flaky than I'd like. I think we might want to stop letting them block us from merging changes.

File: src/internet_identity/src/main.rs [hunk 2]
```diff
@@ -262,7 +322,7 @@ async fn add(user_number: UserNumber, device_data: DeviceData) {
             ));
         }
 
-        entries.push(device_data);
+        entries.push(DeviceDataInternal::from(device_data));
```
Comment (dsarlis) [inline src/internet_identity/src/main.rs:325]: Not sure if we should try to "fix" older devices for this user by setting explicitly to `Some` value for `purpose` and `key_type`. Thoughts?

Comment (kritzcreek) [reply src/internet_identity/src/main.rs:325]: I probably wouldn't bother, given that we're already doing it lazily per account...

**Review by Dfinity-Bjoern — COMMENTED** [review]
Thanks @dsarlis! Looks good to me – I'd like @kritzcreek to have a look as well!

**Review by kritzcreek — APPROVED** [review]
LGTM, thanks!

**Pull request merged**